File: benchmark/android/src/main/kotlin/dev/shreeman/benchmark/BenchmarkPlugin.kt
package dev.shreeman.benchmark

import io.flutter.embedding.engine.plugins.FlutterPlugin
import nitro.benchmark_module.BenchmarkJniBridge
import nitro.nitro_ar_module.NitroArJniBridge

class BenchmarkPlugin : FlutterPlugin {

    companion object {
        init {
            System.loadLibrary("benchmark")
            System.loadLibrary("benchmark_cpp")
            // Required before NitroArJniBridge.registerFactory — its JNI
            // initialize() lives in libnitro_ar.so; without this the load
            // fails with UnsatisfiedLinkError inside onAttachedToEngine,
            // which also kills the MethodChannel registration below it.
            System.loadLibrary("nitro_ar")
        }
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        // registerFactory: one impl per Dart-side instance (multi-instance
        // registry) — the old register(impl) API no longer exists.
        BenchmarkJniBridge.registerFactory({ BenchmarkImpl(binding.applicationContext) }, binding.applicationContext)
        NitroArJniBridge.registerFactory({ NitroArImpl() }, binding.applicationContext)

        val channel = io.flutter.plugin.common.MethodChannel(
            binding.binaryMessenger,
            "dev.shreeman.benchmark/method_channel"
        )
        channel.setMethodCallHandler { call, result ->
            try {
                when (call.method) {
                    "add" -> {
                        val a = call.argument<Double>("a") ?: 0.0
                        val b = call.argument<Double>("b") ?: 0.0
                        result.success(a + b)
                    }
                    "sendLargeBuffer" -> {
                        val buffer = call.arguments as? ByteArray
                        if (buffer != null) {
                            var sum = 0
                            for (i in buffer.indices step 4096) {
                                sum += buffer[i].toInt()
                            }
                        }
                        result.success(buffer?.size ?: 0)
                    }
                    "deviceInfo" -> {
                        // Hardware identity for the benchmark report — a
                        // number without the machine it ran on is not
                        // comparable to anything.
                        result.success(mapOf(
                            "model" to android.os.Build.MODEL,
                            "manufacturer" to android.os.Build.MANUFACTURER,
                            "device" to android.os.Build.DEVICE,
                            "socOs" to "Android ${android.os.Build.VERSION.RELEASE} (SDK ${android.os.Build.VERSION.SDK_INT})",
                        ))
                    }
                    "sievePrimes" -> {
                        // Second reference workload: sieve of Eratosthenes —
                        // identical to src/nitro_workload.h; every tier must
                        // return the same prime count.
                        val limit = call.argument<Int>("limit") ?: 0
                        if (limit < 2) {
                            result.success(0L)
                        } else {
                            val composite = BooleanArray(limit)
                            var count = 0L
                            var i = 2
                            while (i < limit) {
                                if (!composite[i]) {
                                    count++
                                    var j = i.toLong() * i
                                    while (j < limit) {
                                        composite[j.toInt()] = true
                                        j += i
                                    }
                                }
                                i++
                            }
                            result.success(count)
                        }
                    }
                    "hashBuffer" -> {
                        // Reference workload: FNV-1a 64-bit — identical to
                        // src/nitro_workload.h; Long multiplication wraps
                        // mod 2^64, matching C uint64_t bit-for-bit.
                        val data = call.argument<ByteArray>("data") ?: ByteArray(0)
                        val rounds = call.argument<Int>("rounds") ?: 1
                        var hash = -3750763034362895579L // 0xcbf29ce484222325
                        for (r in 0 until rounds) {
                            for (b in data) {
                                hash = hash xor (b.toLong() and 0xFF)
                                hash *= 1099511628211L // 0x100000001b3
                            }
                        }
                        result.success(hash)
                    }
                    else -> result.notImplemented()
                }
            } catch (e: Exception) {
                android.util.Log.e("NitroBenchmark", "MethodChannel Error: ${e.message}", e)
                result.error("ERR", e.message, null)
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {}
}

File: benchmark/macos/Classes/BenchmarkPlugin.swift
import FlutterMacOS
import Foundation

public class BenchmarkPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    NitroArRegistry.register(NitroArModuleImpl())
    BenchmarkRegistry.register(BenchmarkImpl())
    // BenchmarkCpp (NativeImpl.cpp) auto-registers via __attribute__((constructor))
    // in HybridBenchmarkCpp.cpp on library load — no Swift-side registration needed.

    let channel = FlutterMethodChannel(name: "dev.shreeman.benchmark/method_channel", binaryMessenger: registrar.messenger)
    registrar.addMethodCallDelegate(BenchmarkPlugin(), channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    if call.method == "add" {
      let args = call.arguments as? [String: Any]
      let a = args?["a"] as? Double ?? 0.0
      let b = args?["b"] as? Double ?? 0.0
      result(a + b)
    } else if call.method == "sendLargeBuffer" {
      guard let bufferArray = call.arguments as? FlutterStandardTypedData else {
        print("❌ [NitroBenchmark] MethodChannel Error: Invalid buffer data")
        result(FlutterError(code: "ERR", message: "Invalid buffer", details: nil))
        return
      }
      result(Int64(bufferArray.data.count))
    } else if call.method == "deviceInfo" {
      // Hardware identity for the benchmark report.
      var size = 0
      sysctlbyname("hw.model", nil, &size, nil, 0)
      var model = [CChar](repeating: 0, count: size)
      sysctlbyname("hw.model", &model, &size, nil, 0)
      let os = ProcessInfo.processInfo.operatingSystemVersionString
      result([
        "model": String(cString: model),
        "socOs": "macOS \(os)",
      ])
    } else if call.method == "sievePrimes" {
      // Second reference workload: sieve of Eratosthenes — identical to
      // src/nitro_workload.h; every tier must return the same prime count.
      let args = call.arguments as? [String: Any]
      let limit = args?["limit"] as? Int ?? 0
      if limit < 2 { result(Int64(0)); return }
      var composite = [Bool](repeating: false, count: limit)
      var count: Int64 = 0
      var i = 2
      while i < limit {
        if !composite[i] {
          count += 1
          var j = i * i
          while j < limit {
            composite[j] = true
            j += i
          }
        }
        i += 1
      }
      result(count)
    } else if call.method == "hashBuffer" {
      // Reference workload: FNV-1a 64-bit — identical to
      // src/nitro_workload.h; &* wraps mod 2^64, matching C uint64_t.
      let args = call.arguments as? [String: Any]
      let data = (args?["data"] as? FlutterStandardTypedData)?.data ?? Data()
      let rounds = args?["rounds"] as? Int ?? 1
      var hash: UInt64 = 0xcbf29ce484222325
      data.withUnsafeBytes { (ptr: UnsafeRawBufferPointer) in
        let bytes = ptr.bindMemory(to: UInt8.self)
        for _ in 0..<rounds {
          for i in 0..<bytes.count {
            hash ^= UInt64(bytes[i])
            hash = hash &* 0x100000001b3
          }
        }
      }
      result(Int64(bitPattern: hash))
    } else {
      result(FlutterMethodNotImplemented)
    }
  }
}
